plugins/FrequencyShifter/FrequencyShifterEffect.cpp

```diff
@@ -0,0 +1,294 @@
+				
+				float cosP = c;
+				float sinP = s;
+
+				if (doHarm)
+				{
+					// arbitrary distortion function, crossfaded with original signal
+					const float xc = std::clamp(harmFactor * cosP, -3.f, 3.f);
+					const float xs = std::clamp(harmFactor * sinP, -3.f, 3.f);
+					const float xc2 = xc * xc;
+					const float xs2 = xs * xs;
+					const float tc = xc * (27.f + xc2) / (27.f + 9.f * xc2);
+					const float ts = xs * (27.f + xs2) / (27.f + 9.f * xs2);
+					cosP = std::lerp(cosP, tc * harmDiv, harmonics);
+					sinP = std::lerp(sinP, ts * harmDiv, harmonics);
+				}
+
+				float analytic1[2];
+				m_hilbert1.processReal(fxIn[ch], ch, analytic1);
+
+				const float oscR = cosP;
+				const float oscI = sinP * invRing;
+
+				const float modR = analytic1[0] * oscR - analytic1[1] * oscI;
+				const float modI = analytic1[0] * oscI + analytic1[1] * oscR;
+
+				float shiftedR;
+				float shiftedI;
+
+				if (antireflect)
+				{
+					float mod[2] = {modR, modI};
+					float analytic2[2];
+					m_hilbert2.processComplex(mod, ch, analytic2);
+					shiftedR = analytic2[0] * 0.5f;
+					shiftedI = analytic2[1] * 0.5f;
+				}
+				else
+				{
+					shiftedR = modR;
+					shiftedI = modI;
+				}
+
+				m_phase[ch] += dPh[ch];
+				out[ch] = shiftedR + ring * shiftedI;
+			}
+
+			outL = out[0];
+			outR = out[1];
+		}
+
+		float delayInL = outL + (parallelFB ? inL : 0.f);
+		float delayInR = outR + (parallelFB ? inR : 0.f);
+
+		// saturate feedback loop to ensure it doesn't explode
+		constexpr float FbSaturation = 16.f;
+		delayInL = (FbSaturation * delayInL) / (FbSaturation + std::fabs(delayInL));
+		delayInR = (FbSaturation * delayInR) / (FbSaturation + std::fabs(delayInR));
+
+		// 1-pole lowpass in feedback loop
+		m_dampState[0] = (1.f - dampCoeff) * delayInL + dampCoeff * m_dampState[0];
+		m_dampState[1] = (1.f - dampCoeff) * delayInR + dampCoeff * m_dampState[1];
+		m_ringBuf[m_writeIndex][0] = m_dampState[0];
+		m_ringBuf[m_writeIndex][1] = m_dampState[1];
+
+		// 1-pole lowpass on entire signal
+		m_toneState[0] = (1.f - toneCoeff) * outL + toneCoeff * m_toneState[0];
+		m_toneState[1] = (1.f - toneCoeff) * outR + toneCoeff * m_toneState[1];
+		outL = m_toneState[0];
+		outR = m_toneState[1];
+
+		if (routeAdd)
+		{
+			buf[i][0] = inL + mix * outL;
+			buf[i][1] = inR + mix * outR;
+		}
+		else
+		{
+			const float dry = 1.f - mix;
+			buf[i][0] = dry * inL + mix * outL;
+			buf[i][1] = dry * inR + mix * outR;
+		}
+	}
+
+	return ProcessStatus::ContinueIfNotQuiet;
+}
+
+void FrequencyShifterEffect::updateSampleRate()
+{
+	m_sampleRate = Engine::audioEngine()->outputSampleRate();
+
+	m_hilbert1 = HilbertIIRFloat<2>(m_sampleRate, 2.0f);
+	m_hilbert2 = HilbertIIRFloat<2>(m_sampleRate, 2.0f);
+}
+
+extern "C"
+{
+PLUGIN_EXPORT Plugin* lmms_plugin_main(Model* parent, void* data)
+{
+	return new FrequencyShifterEffect(parent, static_cast<const Plugin::Descriptor::SubPluginFeatures::Key*>(data));
+}
+}
```

Comment:
`lmms_plugin_main()` should be moved into the `extern "C"` block at the top; there's no reason new plugins need to perpetuate the weird pattern of having two of these `extern "C"` blocks.